src/test/java/htsjdk/samtools/SAMRecordUnitTest.java

```diff
@@ -1169,4 +1170,86 @@ public void testReadStringGetReadLength(final String readBases, final int readLe
     public void testHasAttribute(final SAMRecord samRecord, final String tag, final boolean expectedHasAttribute) {
         Assert.assertEquals(samRecord.hasAttribute(tag), expectedHasAttribute);
     }
+
+    @Test
+    public void test_setAttribute_empty_array() {
+        final SAMFileHeader header = new SAMFileHeader();
+        final String arrayTag = "XA";
+        final SAMRecord record = new SAMRecord(header);
+        Assert.assertNull(record.getStringAttribute(arrayTag));
+        record.setAttribute(arrayTag, new int[0]);
+        Assert.assertNotNull(record.getSignedIntArrayAttribute(arrayTag));
+        Assert.assertEquals(record.getSignedIntArrayAttribute(arrayTag),new int[0]);
+        record.getSAMString();
+        Assert.assertEquals(record.getAttribute(arrayTag), new char[0]);
+        record.setAttribute(arrayTag, null);
+        Assert.assertNull(record.getStringAttribute(arrayTag));
+
+    }
+
+    @DataProvider
+    public Object[][] getEmptyArrays(){
+        final String BAM = BamFileIoUtils.BAM_FILE_EXTENSION;
```

Comment:
I think this can be replaced with streaming operations, e.g.
```
    public Iterator<Object[]> getEmptyArrays() {
        return Stream.of(BamFileIoUtils.BAM_FILE_EXTENSION, IOUtil.SAM_FILE_EXTENSION, CramIO.CRAM_FILE_EXTENSION)
                .flatMap(ext -> Stream.of(new int[0], new short[0], new byte[0], new float[0])
                        .map(array -> new Object[]{array, array.getClass(), ext})).iterator();
    }
```

Reply:
personally, I do not find that to be more clear...less line? yes...but not more clear.

Reply:
I agree with the sentiment but not this particular stream operation.  Instead I've added a new cartesianProduct method which allows you to combine dataproviders in a nice way.